File: packages/urdu-js/tests/normalization/normalize-characters.test.ts
import { describe, it, expect } from 'vitest'
import { normalizeCharacters } from '../../src/normalization/normalize-characters.js'

describe('normalizeCharacters', () => {
  it('empty string returns empty', () => expect(normalizeCharacters('')).toBe(''))
  it('Arabic ye → Urdu ye', () => expect(normalizeCharacters('\u064a')).toBe('\u06cc'))
  it('Arabic kaf → Urdu kaf', () => expect(normalizeCharacters('\u0643')).toBe('\u06a9'))
  it('Arabic heh → Urdu heh goal', () => expect(normalizeCharacters('\u0647')).toBe('\u06c1'))
  it('converts all three in a word', () => {
    // ہے written with Arabic chars: ه (U+0647) + ي (U+064A)
    expect(normalizeCharacters('\u0647\u064a')).toBe('\u06c1\u06cc')
  })
  it('correct Urdu chars are unchanged', () => {
    const word = 'پاکستان'
    expect(normalizeCharacters(word)).toBe(word)
  })
  it('mixed Arabic+Urdu word: کتاب with Arabic kaf', () => {
    // کتاب with U+0643 instead of U+06A9
    expect(normalizeCharacters('\u0643\u062a\u0627\u0628')).toBe('\u06a9\u062a\u0627\u0628')
  })
  it('preserves non-Arabic characters', () => {
    expect(normalizeCharacters('Hello علم')).toBe('Hello علم')
  })
  it('long text processes under 100ms', () => {
    const long = '\u0647\u064a'.repeat(25000)
    const t0 = performance.now()
    normalizeCharacters(long)
    expect(performance.now() - t0).toBeLessThan(100)
  })
})

File: packages/urdu-dart/test/normalization/normalize_characters_test.dart
import 'package:test/test.dart';
import 'package:urdu_tools/urdu_tools.dart';

void main() {
  group('normalizeCharacters', () {
    test('empty string returns empty',
        () => expect(normalizeCharacters(''), equals('')));
    test('Arabic ye -> Urdu ye',
        () => expect(normalizeCharacters('\u064a'), equals('\u06cc')));
    test('Arabic kaf -> Urdu kaf',
        () => expect(normalizeCharacters('\u0643'), equals('\u06a9')));
    test('Arabic heh -> Urdu heh goal',
        () => expect(normalizeCharacters('\u0647'), equals('\u06c1')));
    test('converts all three in a word', () {
      expect(normalizeCharacters('\u0647\u064a'), equals('\u06c1\u06cc'));
    });
    test('correct Urdu chars are unchanged', () {
      final word = 'پاکستان';
      expect(normalizeCharacters(word), equals(word));
    });
    test('mixed Arabic+Urdu word: کتاب with Arabic kaf', () {
      expect(normalizeCharacters('\u0643\u062a\u0627\u0628'),
          equals('\u06a9\u062a\u0627\u0628'));
    });
    test('preserves non-Arabic characters', () {
      expect(normalizeCharacters('Hello رحم'), equals('Hello رحم'));
    });
  });
}
